File: app/src/main/java/torille/fi/lurkforreddit/utils/TextHelper.java
package torille.fi.lurkforreddit.utils;

import android.text.Html;
import android.text.Spanned;
import android.text.format.DateUtils;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentChild;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentResponse;
import torille.fi.lurkforreddit.data.models.jsonResponses.MultiredditListing;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostDetails;
import torille.fi.lurkforreddit.data.models.jsonResponses.PostResponse;
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditChildren;
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditResponse;
import torille.fi.lurkforreddit.data.models.view.Comment;
import torille.fi.lurkforreddit.data.models.view.Post;
import torille.fi.lurkforreddit.data.models.view.SearchResult;
import torille.fi.lurkforreddit.data.models.view.Subreddit;

/**
 * Contains utility classes for parsing text
 */

public class TextHelper {
    private TextHelper() {
    }

    public static String getLastFourChars(String url) {
        return url.substring(url.length() - 4);
    }

    @SuppressWarnings("deprecation")
    private static Spanned fromHtml(final String html) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            return Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY);
        } else {
            return Html.fromHtml(html);
        }
    }

    private static String formatScore(int score) {
        final String value = String.valueOf(score);
        if (score < 1000) {
            return value;
        } else if (score < 10000) {
            return value.charAt(0) + "." + value.charAt(1) + "k";
        } else if (score < 100000) {
            return value.charAt(0) + value.charAt(1) + "k";
        } else if (score < 10000000) {
            return value.charAt(0) + value.charAt(1) + value.charAt(2) + "k";
        } else {
            return value;
        }
    }

    public static List<Comment> flatten(List<CommentChild> list, int level) {
        List<Comment> flatComments = new ArrayList<>(list.size());
        flatComments = flattenHelper(list, flatComments, level);
        return flatComments;
    }

    public static List<Comment> flattenAdditionalComments(List<CommentChild> list, int level) {
        List<Comment> additionalComments = flatten(list, level);
        for (int i = 0, size = additionalComments.size(); i < size; i++) {
            for (int j = 0, nestedSize = additionalComments.size(); j < nestedSize; j++) {
                if (additionalComments.get(i).name().equals(additionalComments.get(j).parentId())) {
                    additionalComments.set(j, additionalComments.get(j).withLevel(level + 1));
                }
            }
        }
        return additionalComments;
    }

    private static Comment formatCommentData(CommentResponse commentResponse, int level) {
        CharSequence commentText = "";
        CharSequence author = "";
        Comment.kind kind = Comment.kind.MORE;

        if (commentResponse.bodyHtml() != null) {
            kind = Comment.kind.DEFAULT;
            commentText = formatTextToHtml(commentResponse.bodyHtml());
        } else if (commentResponse.id().equals("_")) {
            commentText = "Continue this thread";
        } else if (commentResponse.children() != null) {
            commentText = "Load more comments (" + commentResponse.count() + ")";
        }

        String responseAuthor = commentResponse.author();
        if (responseAuthor != null && commentResponse.stickied()) {
            author = "<font color='#64FFDA'> Sticky post </font>" + responseAuthor;
        } else if (responseAuthor != null) {
            author = responseAuthor;
        }

        author = fromHtml(author + " " + DateUtils.getRelativeTimeSpanString(commentResponse.createdUtc() * 1000, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS));

        if (commentResponse.edited()) {
            author = author + " (edited)";
        }

        String formatScore = formatScore(commentResponse.score());


        return Comment.builder()
                .setId(commentResponse.id())
                .setParentId(commentResponse.parentId())
                .setName(commentResponse.name())
                .setKind(kind)
                .setAuthor(author)
                .setChildCommentIds(commentResponse.children())
                .setCommentLevel(level)
                .setCommentLinkId(commentResponse.linkId())
                .setCommentText(commentText)
                .setFormattedTime("")
                .setFormattedScore(formatScore)
                .setReplies(null)
                .build();
    }

    private static List<Comment> flattenHelper(List<CommentChild> nestedList, List<Comment> flatList, int level) {

        for (int i = 0, size = nestedList.size(); i < size; i++) {

            CommentResponse commentResponse = nestedList.get(i).data();

            Comment comment = formatCommentData(commentResponse, level);
            //check if its a normal comment or "load more comments" comment
            flatList.add(comment);
            if (commentResponse != null) {
                CommentListing commentListing = commentResponse.replies();
                if (commentListing != null && commentListing.commentData() != null) {
                    List<Comment> replyList = flattenHelper(commentListing.commentData().commentChildren(), flatList, level + 1);
                    comment.withReplies(replyList);
                }
            }

        }
        return flatList;
    }


    private static Function<PostDetails, Post> funcFormatPost = new Function<PostDetails, Post>() {
        @Override
        public Post apply(@NonNull PostDetails postDetails) throws Exception {
            String formatScore = TextHelper.formatScore(postDetails.score());

            CharSequence title = TextHelper.fromHtml(postDetails.title());

            // sometimes formatting title can result in empty if it has <----- at start
            // etc
            if (title != null &&
                    title.length() == 0) {
                title = String.valueOf(postDetails.title());
            }

            CharSequence selfText = null;
            if (postDetails.selftextHtml() != null) {
                selfText = formatTextToHtml(postDetails.selftextHtml());
            }

            CharSequence flair = "";
            if (postDetails.stickied()) {
                flair = TextHelper.fromHtml("Stickied");
            }
            if (postDetails.linkFlairText() != null) {
                flair = TextHelper.fromHtml(flair + " " + postDetails.linkFlairText());
            }
            String previewImageUrl;
            String thumbnail = postDetails.thumbnail();
            switch (thumbnail) {
                case "default":
                case "self":
                case "":
                case "spoiler":
                case "image":
                    previewImageUrl = "";
                    break;
                case "nsfw":
                    flair = TextHelper.fromHtml("<font color='#FF1744'>NSFW </font>" + flair);
                    previewImageUrl = "";
                    break;
                default:
                    previewImageUrl = DisplayHelper.getBestPreviewPicture(postDetails);
            }


        /*if (postDetails.isOver18()) {
            title = TextHelper.fromHtml("<font color='#FF1744'> NSFW </font>" + postDetails.title());
        }*/

            String numberOfComments = String.valueOf(postDetails.numberOfComments());

            return Post.builder()
                    .setId(postDetails.name())
                    .setThumbnail(thumbnail)
                    .setDomain(postDetails.domain())
                    .setUrl(postDetails.url())
                    .setScore(formatScore)
                    .setFlairText(flair)
                    .setSelfText(selfText)
                    .setIsSelf(postDetails.isSelf())
                    .setNumberOfComments(numberOfComments)
                    .setTitle(title)
                    .setPreviewImage(previewImageUrl)
                    .setPermaLink(postDetails.permalink())
                    .setAuthor(postDetails.author())
                    .setCreatedUtc(postDetails.createdUtc())
                    .build();
        }
    };

    public static Observable<Post> formatPost(Observable<PostResponse> postDetails) {
        return postDetails
                .map(PostResponse::postDetails)
                .map(funcFormatPost);

    }

    public static Observable<Post> formatPostDetails(Observable<PostDetails> postDetails) {
        return postDetails
                .map(funcFormatPost);

    }

    public static Observable<Subreddit> formatMultiToSubreddit(Observable<MultiredditListing> multireddit) {

        return multireddit.map(multiredditListing -> {
            MultiredditListing.Multireddit multireddit1 = multiredditListing.multireddit();
            return Subreddit.builder()
                    .setUrl(multireddit1.pathUrl())
                    .setBannerUrl(null)
                    .setId(multireddit1.name())
                    .setKeyColor(multireddit1.keyColor())
                    .setDisplayName(multireddit1.displayName())
                    .build();
        });

    }

    public static Observable<Subreddit> formatSubreddit(Observable<SubredditChildren> subredditChildren) {

        return subredditChildren
                .map(SubredditChildren::subreddit)
                .map(subredditResponse -> Subreddit.builder()
                        .setUrl(subredditResponse.url())
                        .setBannerUrl(subredditResponse.banner())
                        .setId(subredditResponse.id())
                        .setKeyColor(subredditResponse.keyColor())
                        .setDisplayName(subredditResponse.displayName())
                        .build());
    }

    public static Observable<SearchResult> formatSearchResult(Observable<SubredditChildren> observable) {
        return observable.map(subredditChildren -> {
            Observable<SubredditChildren> subredditChildObservable = Observable.<SubredditChildren>fromArray(subredditChildren);
            return Observable.zip(formatSubreddit(subredditChildObservable),
                    subredditChildObservable,
                    funcFormatSearchResult)
                    .blockingSingle();

        });
    }

    private static BiFunction<Subreddit, SubredditChildren, SearchResult> funcFormatSearchResult = (subreddit, subredditChildren) -> {
        SubredditResponse subredditResponse = subredditChildren.subreddit();

        CharSequence formattedTitle;
        String formattedSubscription;
        CharSequence formattedDescription;

        final String infoText = subredditResponse.subscribers() + " subscribers, Community since "
                + DateUtils.getRelativeTimeSpanString(subredditResponse.createdUtc() * 1000);

        if (subredditResponse.over18()) {
            formattedTitle = TextHelper.fromHtml(subredditResponse.url()
                    + "<font color='#FF1744'> NSFW</font>");
        } else {
            formattedTitle = subredditResponse.url();
        }

        if (subredditResponse.subscribed()) {
            formattedSubscription = "Subscribed";
        } else {
            formattedSubscription = "Not subscribed";
        }
        String descriptionHtml = subredditResponse.descriptionHtml();
        if (descriptionHtml != null && !descriptionHtml.isEmpty()) {
            formattedDescription = formatTextToHtml(descriptionHtml);
        } else {
            formattedDescription = "No description";
        }

        return SearchResult.builder()
                .setTitle(formattedTitle)
                .setDescription(formattedDescription)
                .setInfoText(infoText)
                .setSubscriptionInfo(formattedSubscription)
                .setSubreddit(subreddit)
                .build();
    };

    /**
     * Trims trailing whitespace. Removes any of these characters:
     * 0009, HORIZONTAL TABULATION
     * 000A, LINE FEED
     * 000B, VERTICAL TABULATION
     * 000C, FORM FEED
     * 000D, CARRIAGE RETURN
     * 001C, FILE SEPARATOR
     * 001D, GROUP SEPARATOR
     * 001E, RECORD SEPARATOR
     * 001F, UNIT SEPARATOR
     *
     * @return "" if source is null, otherwise string with all trailing whitespace removed
     */

    private static CharSequence trimTrailingWhitespace(CharSequence source) {

        if (source == null)
            return "";

        int i = source.length();

        // loop back to the first non-whitespace character
        while (--i >= 0 && Character.isWhitespace(source.charAt(i))) {
        }

        return source.subSequence(0, i + 1);
    }

    private static CharSequence formatTextToHtml(String bodyText) {
        if (bodyText == null) {
            return "";
        }
        Spanned htmlText = fromHtml(bodyText);
        return trimTrailingWhitespace(htmlText);
    }
}

File: app/src/main/java/torille/fi/lurkforreddit/utils/TextHelper.kt
package torille.fi.lurkforreddit.utils

import android.text.Html
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.format.DateUtils
import io.reactivex.Flowable
import io.reactivex.functions.Function
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentChild
import torille.fi.lurkforreddit.data.models.jsonResponses.CommentResponse
import torille.fi.lurkforreddit.data.models.jsonResponses.PostDetails
import torille.fi.lurkforreddit.data.models.jsonResponses.SubredditChildren
import torille.fi.lurkforreddit.data.models.view.*

/**
 * Contains utility classes for parsing text
 */

object TextHelper {

    fun getLastFourChars(url: String): String {
        return url.substring(url.length - 4)
    }

    private fun fromHtml(html: String): Spanned {
        return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY)
        } else {
            Html.fromHtml(html)
        }
    }

    private fun formatScore(score: Int): String {
        val value = score.toString()
        return when {
            score < 1000 -> value
            score < 10000 -> value[0] + "." + value[1] + "k"
            score < 100000 -> value[0] + "" + value[1] + "k"
            score < 10000000 -> value[0] + "" + value[1] + "" + value[2] + "k"
            else -> value
        }
    }

    fun flatten(list: List<CommentChild>, commentDepth: Int): List<Comment> {
        return list.flatMap { formatCommentData(it.data, commentDepth) }
    }

    fun flattenAdditionalComments(list: List<CommentChild>, level: Int): List<Comment> {

        val additionalComments = flatten(list, level).toMutableList()

        var i = 0
        val size = additionalComments.size
        while (i < size) {
            var j = 0
            val nestedSize = additionalComments.size
            while (j < nestedSize) {
                if (additionalComments[i].name == additionalComments[j].parentId) {
                    additionalComments[j] = additionalComments[j].copy(commentLevel = level + 1)
                }
                j++
            }
            i++
        }

        return additionalComments.toList()
    }

    private fun formatCommentData(commentData: CommentResponse?, commentDepth: Int): List<Comment> {

        return if (commentData != null) {
            var commentText: CharSequence = ""
            var author: CharSequence = ""
            var kind: kind = kind.MORE

            when {
                commentData.bodyHtml.isNotEmpty() -> {
                    kind = torille.fi.lurkforreddit.data.models.view.kind.DEFAULT
                    commentText = formatTextToHtml(commentData.bodyHtml)
                }
                commentData.id == "_" -> commentText = "Continue this thread"
                commentData.children.isNotEmpty() -> commentText = "Load more comments (" + commentData.count + ")"
            }

            val responseAuthor = commentData.author
            if (responseAuthor.isNotEmpty() && commentData.stickied) {
                author = "<font color='#64FFDA'> Sticky post </font>" + responseAuthor
            } else if (responseAuthor.isNotEmpty()) {
                author = responseAuthor
            }

            author = fromHtml(author.toString() + " " + DateUtils.getRelativeTimeSpanString(commentData.createdUtc * 1000, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS))

            if (commentData.edited) {
                author = author.toString() + " (edited)"
            }

            val formatScore = formatScore(commentData.score)

            val formattedComment = Comment(
                    id = commentData.id,
                    parentId = commentData.parentId,
                    name = commentData.name,
                    kind = kind,
                    author = author,
                    childCommentIds = commentData.children,
                    commentLevel = commentDepth,
                    commentLinkId = commentData.linkId,
                    commentText = commentText,
                    formattedScore = formatScore,
                    formattedTime = "",
                    replies = null
            )

            val listWithOriginalComment = listOf(formattedComment)

            return if (commentData.replies != null) {
                val replies = commentData.replies.commentData.commentChildren
                        .flatMap { formatCommentData(it.data, commentDepth + 1) }

                listWithOriginalComment.plus(replies)
            } else {
                listWithOriginalComment
            }
        } else {
            emptyList()
        }


    }

    val funcFormatPost = Function { postDetails: PostDetails ->
        val formatScore = TextHelper.formatScore(postDetails.score)
        var selfText: CharSequence = ""
        var previewImageUrl = ""
        val thumbnail = postDetails.thumbnail
        var title: CharSequence = TextHelper.fromHtml(postDetails.title)
        var flair: Spanned = SpannableStringBuilder.valueOf("")

        // sometimes formatting title can result in empty if it has <----- at start
        // etc
        if (title.isEmpty()) {
            title = postDetails.title
        }

        if (postDetails.selftextHtml != null) {
            selfText = formatTextToHtml(postDetails.selftextHtml)
        }

        if (postDetails.stickied) {
            flair = TextHelper.fromHtml("Stickied")
        }
        if (postDetails.linkFlairText != null) {
            flair = TextHelper.fromHtml(flair.toString() + " " + postDetails.linkFlairText)
        }

        when (thumbnail) {
        //"default", "self", "", "spoiler", "image" -> previewImageUrl = ""
            "nsfw" -> {
                flair = TextHelper.fromHtml("<font color='#FF1744'>NSFW </font>" + flair)
            }
            else -> previewImageUrl = DisplayHelper.getBestPreviewPicture(postDetails)
        }
        val dashUrl = postDetails.media?.redditVideo?.dashUrl
        val fallbackUrl = postDetails.media?.redditVideo?.fallbackUrl
        val url: String = if (!dashUrl.isNullOrEmpty() && postDetails.media?.redditVideo?.transcodingStatus == "completed") {
            dashUrl!!
        } else if (!fallbackUrl.isNullOrEmpty()) {
            fallbackUrl!!
        } else {
            postDetails.url
        }
        /*if (postDetails.isOver18()) {
        title = TextHelper.fromHtml("<font color='#FF1744'> NSFW </font>" + postDetails.title());
        }*/

        val numberOfComments = postDetails.numberOfComments
        Post(
                id = postDetails.name,
                thumbnail = thumbnail,
                domain = postDetails.domain,
                url = url,
                score = formatScore,
                flairText = flair,
                selfText = selfText,
                isSelf = postDetails.isSelf,
                numberOfComments = numberOfComments.toString(),
                title = title,
                previewImage = previewImageUrl,
                permaLink = postDetails.permalink,
                author = postDetails.author,
                createdUtc = postDetails.createdUtc
        )
    }


    fun formatSubreddit(subredditChildren: Flowable<SubredditChildren>): Flowable<Subreddit> {

        return subredditChildren
                .map { it.subreddit }
                .map { subredditResponse ->
                    // TODO("If subreddit isnt found return something else than empty")
                    if (subredditResponse.id.isNullOrBlank()) {
                        Subreddit()
                    } else {
                        Subreddit(
                                url = subredditResponse.url,
                                bannerUrl = subredditResponse.banner,
                                subId = subredditResponse.id!!,
                                keyColor = subredditResponse.keyColor,
                                displayName = subredditResponse.displayName
                        )
                    }

                }
    }

    fun formatSearchResult(flowable: Flowable<SubredditChildren>): Flowable<SearchResult> {
        return flowable.flatMap { subredditChildren ->
            val subredditChildFlowable = Flowable.fromArray(subredditChildren)
            Flowable.zip<Subreddit, SubredditChildren, SearchResult>(
                    formatSubreddit(subredditChildFlowable),
                    subredditChildFlowable,
                    io.reactivex.functions.BiFunction({ subreddit: Subreddit, subredditChild: SubredditChildren ->

                        val subredditResponse = subredditChild.subreddit
                        val descriptionHtml = subredditResponse.descriptionHtml

                        val formattedTitle: CharSequence
                        val formattedSubscription: String
                        val formattedDescription: CharSequence

                        val infoText = subredditResponse.subscribers.toString() + " subscribers, Community since " + DateUtils.getRelativeTimeSpanString(subredditResponse.createdUtc * 1000)

                        formattedTitle = if (subredditResponse.over18) {
                            TextHelper.fromHtml(subredditResponse.url + "<font color='#FF1744'> NSFW</font>")
                        } else {
                            subredditResponse.url
                        }

                        formattedSubscription = if (subredditResponse.subscribed) {
                            "Subscribed"
                        } else {
                            "Not subscribed"
                        }

                        formattedDescription = if (!descriptionHtml.isNullOrEmpty()) {
                            formatTextToHtml(descriptionHtml)
                        } else {
                            "No description"
                        }

                        SearchResult(
                                title = formattedTitle,
                                description = formattedDescription,
                                infoText = infoText,
                                subscriptionInfo = formattedSubscription,
                                subreddit = subreddit
                        )
                    })
            )

        }
    }

    private fun formatTextToHtml(bodyText: String?): CharSequence {
        if (bodyText == null) {
            return ""
        }
        val htmlText = fromHtml(bodyText)
        return htmlText.trim()
    }
}
